service/frontend/adminHandler.go

```diff
@@ -284,7 +284,7 @@ func (adh *adminHandlerImpl) RemoveTask(
 	scope, sw := adh.startRequestProfile(metrics.AdminRemoveTaskScope)
 	defer sw.Stop()
 
-	if request == nil || request.Type == nil || request.TaskID == nil {
+	if request == nil {
```

Comment:
hmms.  is 0-type a valid type?  (actually, what type is it?  thrift files just leave it as an int....)

Reply:
Yeah, lack of enums makes it harder to understand. You can track it down to this: https://github.com/uber/cadence/blob/master/common/constants.go#L164
We can see here that valid values starts from `2`. Maybe update this check to reject everything under `2`?

Also for proto, I have introduced enum for those: https://github.com/uber/cadence/blob/master/proto/internal/uber/cadence/shared/v1/queue.proto#L27

Reply:
yea, yay improved RPC APIs :)

I'd prefer a check of some kind, yeah.  `>=2` is strictly accurate and might lead to discovering the right type too, but `!=0` might be enough - it at least identifies "did not provide -> default 0" as an invalid value, which is probably good enough until we switch to better types.

Reply:
Actually, I have reverted pointer drop on those fields. As it is mapped to enum value in proto, it also contains `_INVALID` value for it. So we can map nil to `_INVALID` just like for all other enums, making more consistent.

Reply:
Regarding all those checks. I have a rough idea how it should be done. Since we now have pure internal types, it opens quite a few doors for improvement. Once such idea - internal type validation functions. Those types could have:
```
func (t SomeInternalType) Validate() error {
  if t.Field1 <= 0 {
    return Error1
  }
  if t.Field2 == nil {
    return Error2
  }
  etc...
}
```

Then bigger types such as all `Request`s could call `Validate()` of all composed types, like:
```
func (t *SomeFrontendRequest) Validate() error {
  var errors error
  if t == nil {
    errors = multierr.Append(errors, NilRequestError)
  }
  if err := t.Field1.Validate(); err != nil {
    errors = multierr.Append(errors, err)
  }
  if err := t.Field2.Validate(); err != nil {
    errors = multierr.Append(errors, err)
  }
  return errors
}
```

All validation logic would be in once place and easily testable. Then in each handler, we could simple call this `Validate` at the very beginning of a call:
```
func (h *Handler) SomeRPC(request *SomeRequest) (*SomeResponse, error) {
  if err := request.Validate(); err != nil {
    return nil, err
  }
}
```
With this setup we would not need to worry about invalid values possible somewhere down the line. It would be consistent, testable and isolated in once place.

Of course this is for another (not small) task, but that is the direction that I would like to eventually proceed.